### ISSUE TITLE
[Scann] Does example.py uses loss function from the paper?

### ISSUE THREAD

**Issue by YJYJLee** [google-research/google-research#546]
I'm following the code flow by running example.py and defining search as below,
`searcher = scann.ScannBuilder(normalized_dataset, 10, "dot_product").tree(
    num_leaves=2000, num_leaves_to_search=100, training_sample_size=250000).score_ah(2, anisotropic_quantization_threshold=0.2).reorder(100).create_pybind()`

I was expecting to see the part where it trains codewords with the loss function that was proposed in the paper, but I only see training codewords with GenericKmeans (original kmeans-clustering algorithm). 

Does the example.py uses loss function from the paper? If not, how can I change the searcher definition to make it use the loss function from the paper?

**Comment by sammymax** [comment]
The `anisotropic_quantization_threshold=0.2` portion results in the loss function from the paper being used. Removing that value or setting it to NaN will change ScaNN to use standard reconstruction loss.

**Comment by YJYJLee** [comment]
@sammymax  thanks for the explanation, I followed the code and have some few questions. 
1)  I found that `anisotropic_quantization_threshold=0.2` is being used when training pq codebook at hashes/internal/asymmetric_hashing_impl.cc:189. 
`if (!std::isnan(opts.config().noise_shaping_threshold()) &&
      opts.config().use_noise_shaped_training())`
Specifically at this line, it checks whether the threshold was given as parameters and if `use_noise_shaped_training` is set to True. But I checked that use_noise_shaped_trainig's default value is false (refering ./scann/proto/hash.proto) and its value does not change. In summary, above if clause is never true, so it never enters to function ComputeParallelCostMultiplier. So whatever value I give to `anisotropic_quantization_threshold`, parallel_cost_multiplier value is always 1. 
2) Assuming 1) problem is resolved, it still calls 
`ComputeParallelCostMultiplier(opts.config().noise_shaping_threshold(), 1.0, dataset.dimensionality())` at hashes/internal/asymmetric_hashing_impl.cc:191. Refering to the paper, 1.0 should be the norm of the database. It will specially work with glove since we normalized its norms, but for general MIPS problems, norms of the database are not constant. How should we fix it to support dataset with various norms?   
3) How do we set use_noise_shaped_training to True? I can just fix ./scann/proto/hash.proto, but is there any way to set this value via paremter?

I will be very helpful if you answer above questions!

**Comment by vsippola** [comment]
For 2) If they do not have it implemented you will have to calculate/save each heuristic weight value (eq 3 in the paper) and then edit the loss function so it is calculated correctly.

I would assume that "use_noise_shaped_training" is supposed to be set as the noise_shaping_threshold is in scann_builder.py.. however looking at the code that doesn't seem to be the case

**Comment by vsippola** [comment]
Experimenting with changing scann_builder.py causes the runtime of the residual quantitation to explode. ~50x slower.

3283 seconds with the value set to True vs 68 seconds with the value set to False

so I assume that it is correctly doing anisotropic hashing instead of plain k-means




-Assuming- changing this parameter gives the correct behaviour (I have not checked this, just observed runtime differences) you can fix this in a couple ways. One by editing 

scann_builder.py's section

@_factory_decorator("score_ah")


to set a variable

noise_shaped_training = not(math.isnan(anisotropic_quantization_threshold)) and (0 <= anisotropic_quantization_threshold)



Then add in the return section of @_factory_decorator("score_ah")

use_noise_shaped_training: {noise_shaped_training}





(pardon any syntax errors)

**Comment by liql2007** [comment]
I found the same problem. so scann default not use "anisotropic hashing/quantization".

**Comment by liql2007** [comment]
Assuming 1) problem is resolved, but `gmm.GenericKmeans(...)` alway set parameter `spherical` as `false`. The new center points of each k-means iteration are not normalized, this will cause the k-means clustering is imbalance and not converged.

I think `RecomputeCentroidsWithParallelCostMultiplier(...)` should support parameter `spherical` and normalize the centers. 

@sammymax @YJYJLee

**Comment by yohnung** [comment]
I found the same problem, which is changing `anisotropic_quantization_threshold` from 0.001 to 0.4 results in the same recall-rate, and i have no idea how to fix it. It would be very helpful if the author or contributor can reply.
